synapse/storage/background_updates.py

```diff
@@ -740,6 +818,169 @@ def create_index_sqlite(conn: Connection) -> None:
         logger.info("Adding index %s to %s", index_name, table)
         await self.db_pool.runWithConnection(runner)
 
+    def register_background_validate_constraint_and_delete_rows(
+        self,
+        update_name: str,
+        table: str,
+        constraint_name: str,
+        constraint: Constraint,
+        unique_columns: Sequence[str],
+    ) -> None:
+        """Helper for store classes to do a background validate constraint, and
+        delete rows that do not pass the constraint check.
+
```

Comment:
At first I was confused why we could delete rows without losing data. But we're defining one table to be the source of truth and if there is a row in another table that doesn't have a corresponding row in the source of truth then it's just leftover data that we forgot to delete when we deleted from the source of truth table.

It would be nice to explain the nuance here imo.

Reply:
I've added a note about double checking that we're OK with deleting rows in `register_background_validate_constraint_and_delete_rows`